src/sonic-config-engine/portconfig.py

```diff
@@ -172,85 +172,152 @@ def parse_port_config_file(port_config_file):
+        def _speed_to_int(cls, speed):
+            try:
+                if speed.endswith('G'):
+                    return int(speed.replace('G', '')) * 1000
+
+                return int(speed)
+            except ValueError:
+                raise RuntimeError("Unsupported speed format '{}'".format(speed))
+
+        def _parse_supported_speed(self, speed):
+            if not speed:
+                return
+
+            if not speed.startswith('[') and not speed.endswith(']'):
+                raise RuntimeError("Unsupported port breakout format!")
+
+            for s in speed[1:-1].split(','):
+                self.supported_speed.add(self._speed_to_int(s.strip()))
+
+        def _parse_num_assigned_lanes(self, num_assigned_lanes):
+            if not num_assigned_lanes:
+                return
+
+            if isinstance(num_assigned_lanes, int):
+                return num_assigned_lanes
+
+            if not num_assigned_lanes.startswith('(') and not num_assigned_lanes.endswith(')'):
+                raise RuntimeError("Unsupported port breakout format!")
+
+            return int(num_assigned_lanes[1:-1])
+
+        def __eq__(self, other):
+            if isinstance(other, BreakoutCfg.BreakoutModeEntry):
+                if self.num_ports != other.num_ports:
+                    return False
+                if self.supported_speed != other.supported_speed:
+                    return False
+                if self.num_assigned_lanes != other.num_assigned_lanes:
+                    return False
+                return True
             else:
-                raise Exception('Regex return for speed is None...')
+                return False
 
-            ports[intf_name]['index'] = index.split(",")[alias_position]
+        def __ne__(self, other):
+            return not self == other
 
-            parent_intf_id += step
-            alias_position += 1
-            lanes_start += step
+        def __hash__(self):
+            return hash((self.num_ports, tuple(self.supported_speed), self.num_assigned_lanes))
+
+    def __init__(self, name, bmode, properties):
+        self._interface_base_id = int(name.replace(PORT_STR, ''))
+        self._properties = properties
+        self._lanes = properties ['lanes'].split(',')
+        self._indexes = properties ['index'].split(',')
+        self._breakout_mode_entry = self._str_to_entries(bmode)
+        self._breakout_capabilities = None
+
+        # Find specified breakout mode in port breakout mode capabilities
+        for supported_mode in self._properties['breakout_modes']:
+            if self._breakout_mode_entry == self._str_to_entries(supported_mode):
+                self._breakout_capabilities = self._properties['breakout_modes'][supported_mode]
+                break
+
+        if not self._breakout_capabilities:
+            raise RuntimeError("Unsupported breakout mode {}!".format(bmode))
+
+    def _re_group_to_entry(self, group):
+        if len(group) != 6:
```

Comment:
can we make `6` as global var in the beginning of this file and use it here and also in `BRKOUT_PATTERN`?

Reply:
I added a new `BRKOUT_PATTERN_GROUPS=6` global variable to replace the magic number used here.
However, I didn't change the value used in `BRKOUT_PATTERN` because it has a different meaning.
Here be indicating how many groups should have as a result of regexp matching.
In `BRKOUT_PATTERN` pattern `6` indicates up to how many digits can be in each group.
It's just a coincidence that these values are the same :)

Reply:
Aah, makes sense! 👍